packages/repository-json-schema/src/build-schema.ts

```diff
@@ -186,7 +200,18 @@ export function modelToJsonSchema(ctor: Function): JSONSchema {
       continue;
     }
 
-    const propSchema = getJsonSchema(referenceType);
+    // Break to avoid the circular reference.
+    // If the model is already visited in the call stack, it implies one or more
+    // of its referenced properties refer back to it.
+
+    if (isVisited) break;
+
+    // Use object assign to avoid polluting the original `options`.
+    const getJsonSchemaOptions = Object.assign({}, options);
+    getJsonSchemaOptions.visited = getJsonSchemaOptions.visited || [];
+    getJsonSchemaOptions.visited.push(result.title!);
+
+    const propSchema = getJsonSchema(referenceType, getJsonSchemaOptions);
 
     if (propSchema && Object.keys(propSchema).length > 0) {
       result.definitions = result.definitions || {};
```

Comment:
One idea: since we are effectively storing the schema of visited models in `result.definitions`; if we need to access schema of a model that has been already visited, we can perhaps obtain it from those definitions object? Or maybe we should remove `options.visited` and pass the `definitions` key-value map in options instead?

Reply:
> One idea: since we are effectively storing the schema of visited models in result.definitions

Hmm, do you mean storing the schema of the current model in the `definitions`?

Like 
```
{
        title: 'Category',
        properties: {
          name: {
            type: 'string'
          },
        },
        definitions: {
          Category: {
            title: 'Category',
            properties: {
               name: {
                 type: 'string'
               },
            },
          },
        },
}
```

Reply:
> Hmm, do you mean storing the schema of the current model in the `definitions`?

No, only schema of models used by properties (i.e. `Product` when emitting schema for `Category`).